### PR TITLE
Fix normalising smooth cubic Bézier curves without a previous curve

### LINKED ISSUE #11: Problems with normalising smooth cubic Bézier curves
**Edit:** I investigated a bit further and noticed the problem occurs on normalising not on reversing

I've noticed some problems ~~reversing~~ normalising certain paths with smooth cubic Bézier curves (s-Command).

I've prepared two examples to illustrate the problem:

Example 1:
`M249.834 243.462s120.037 20.777 149.951 26.898l129.125 22.261`
Screenshot 2022-09-13 at 12 57 36

Normalised:
`M 249.834 243.462 C 499.668 486.924 369.871 264.239 399.78499999999997 270.36 L 528.91 292.62100000000004`
Screenshot 2022-09-13 at 12 57 19

Example 2:
`M259.238 310.863s7.636.062 8.204-3.154c0 0 1.451-4.514-9.593-5.07 0 0-5.932-.186-4.292-3.278 0 0 .695-1.484 2.651-1.793`
Screenshot 2022-09-13 at 13 09 47

Normalised:
`M 259.238 310.863 C 518.476 621.726 266.874 310.925 267.442 307.709 C 267.442 307.709 268.89300000000003 303.195 257.849 302.639 C 257.849 302.639 251.917 302.45300000000003 253.557 299.361 C 253.557 299.361 254.25199999999998 297.877 256.20799999999997 297.568`
Screenshot 2022-09-13 at 13 09 23

### PR DESCRIPTION
**Pull request by Eschon**
Fixes #11 

choose first control point for smooth cubic bézier curves based on the previous command

if the previous command was another bézier curve use the last control point reflected over the current point otherwise use the current point

### CODE REVIEW

File: reverse.js [hunk 1]
```diff
@@ -34,6 +34,7 @@
         instruction,
         op,
         lop,
+        preop,
```
Comment (Pomax) [inline reverse.js]: let's call this `prevop` so that it's clearer what its role is.

File: reverse.js [hunk 2]
```diff
@@ -194,9 +195,16 @@
...
+          if (["s", "c"].includes(preop)) {
+            // reflect previous cx2/cy2 over x/y
+            cx = x + (x-cx2);
+            cy = y + (y-cy2);
+          }
+          else {
+            cx = x;
+            cy = y;
+          }
+
```
Comment (Pomax) [inline reverse.js]: small tweak, but note that in both cases we assign x/y:
```suggestion
          cx = x;
          cy = y;
          if (["s", "c"].includes(prevop)) {
            cx += (x-cx2);
            cy += (y-cy2);
          }
```

File: test.js [hunk 3]
```diff
@@ -75,3 +75,17 @@ prenormalized = "M 0 0 L 100 0 L 100 100 L 20 100 Z";
...
+normalized = normalize(path);
+if (normalized !== prenormalized) fail(`smooth cubic Bézier without previous Bézier (normalize): ${normalized}`);
+if (prenormalized !== reverse(reverse(path))) fail('smooth cubic Bézier without previous Bézier');
+
+
+path = "M 110,90 c 20,0 15,-80 40,-80 s 20,80 40,80";
+prenormalized = "M 110 90 C 130 90 125 10 150 10 C 175 10 170 90 190 90";
+normalized = normalize(path);
+if (normalized !== prenormalized) fail(`smooth cubic Bézier with previous Bézier (normalize): ${normalized}`);
+if (prenormalized !== reverse(reverse(path))) fail('smooth cubic Bézier with previous Bézier');
```
Comment (Pomax) [inline test.js]: a silly github thing, but this is missing a newline at the end.

**Review by Pomax — CHANGES_REQUESTED** [review]

File: reverse.js (hunk 1 above)
Comment (Eschon) [reply reverse.js]: Ok, done

File: test.js (hunk 3 above)
Comment (Eschon) [reply test.js]: My bad, added one

File: reverse.js [hunk 4]
```diff
@@ -195,16 +195,12 @@
       }
       else if(lop === "s") {
         for (a = 0; a < alen; a += 4) {
-          if (["s", "c"].includes(preop)) {
-            // reflect previous cx2/cy2 over x/y
-            cx = x + (x-cx2);
-            cy = y + (y-cy2);
+          cx = x;
+          cy = y;
+          if (["s", "c"].includes(prevop)) {
```
Comment (Pomax) [inline reverse.js]:
```suggestion
          if (["s", "c"].indexOf(prevop) > -1) {
```

**Comment by Pomax** [comment]
@Eschon one more thought: this probably also applies to the `t` operator for quadratic curves, would it be too much to ask you to add the same changes there with a quadratic test case?

**Comment by Eschon** [comment]
@Pomax sure no problem. I've adapted the fix for quadratic curves

File: reverse.js [hunk 5]
```diff
@@ -157,9 +157,16 @@
...
+          if (["t", "q"].indexOf(prevop) > -1) {
+            // reflect previous cx/cy over x/y
+            cx = x + (x-cx);
+            cy = y + (y-cy);
+          }
+          else {
+            cx = x;
+            cy = y;
+            prevop = lop;
+          }
```
Comment (Pomax) [inline reverse.js]: Let's use the same code as for s/c though (universal x/y assignment with conditional increment)

File: test.js [hunk 6]
```diff
@@ -89,3 +89,17 @@ prenormalized = "M 110 90 C 130 90 125 10 150 10 C 175 10 170 90 190 90";
...
+normalized = normalize(path);
+if (normalized !== prenormalized) fail(`quadratic Bézier without previous Bézier (normalize): ${normalized}`);
+if (prenormalized !== reverse(reverse(path))) fail('quadratic Bézier without previous Bézier');
+
+
+path = "M 10,50 Q 25,25 40,50 t 30,0 30,0 30,0 30,0 30,0"
+prenormalized = "M 10 50 Q 25 25 40 50 Q 55 75 70 50 Q 85 25 100 50 Q 115 75 130 50 Q 145 25 160 50 Q 175 75 190 50";
+normalized = normalize(path);
+if (normalized !== prenormalized) fail(`quadratic Bézier with previous Bézier (normalize): ${normalized}`);
+if (prenormalized !== reverse(reverse(path))) fail('quadratic Bézier with previous Bézier');
```
Comment (Pomax) [inline test.js]: really wish github just ignored this =D

**Review by Pomax — CHANGES_REQUESTED** [review]

File: reverse.js (hunk 5 above)
Comment (Eschon) [reply reverse.js]: Doesn't work in that case. With s/c we reflect cx2/cy2 here we have cx/cy. If we assign them first it breaks

File: test.js (hunk 6 above)
Comment (Eschon) [reply test.js]: my bad, fixed it and also fixed my editor config to add it by default  😅

File: reverse.js (hunk 5 above)
Comment (Pomax) [reply reverse.js]: d'uh, yes of course.

File: reverse.js [hunk 7]
```diff
@@ -156,9 +157,16 @@
...
-          cy = y + (y-cy);
+          if (["t", "q"].indexOf(prevop) > -1) {
+            // reflect previous cx/cy over x/y
+            cx = x + (x-cx);
+            cy = y + (y-cy);
+          }
+          else {
+            cx = x;
+            cy = y;
+            prevop = lop;
```
Comment (Pomax) [inline reverse.js]: We probably want to move this assignment out of the `else`?

**Review by Pomax — CHANGES_REQUESTED** [review]

File: reverse.js (hunk 7 above)
Comment (Eschon) [reply reverse.js]: You're right. Functionally it doesn't really make a difference, I just put it there to make it work.
But it's cleaner to move it out of there

**Review by Pomax — APPROVED** [review]
Awesome, let's get this in!

**Comment by Pomax** [comment]
v1.5.0 released

**Pull request merged**